### ISSUE TITLE
Setting batch_type to 'single_image'

### ISSUE THREAD

**Issue by joanncqn** [hjxwhy/mipnerf_pl#7]
Hi, 

I'm trying to train a model with the batch_type set to 'single_image', but it seems like this part hasn't been implemented so far.
I tried to set shuffle in dataloader to be False, so that each batch has rays only coming from one image. However, it doesn't seem to work as the novel view are just white images.

Any clue what might be a right implementation for 'single_image' mode?
Appreciate any input you might have.

**Comment by hjxwhy** [comment]
Hi, 
'single_image' mode only can be used for 'val' now. If you have RAM limitations I think you can downscale image size https://github.com/hjxwhy/mipnerf_pl/blob/bd88427c206e113d9d6e9d8449d1b8c21d4a8335/datasets/datasets.py#L194. for single scale mode.

**Comment by joanncqn** [comment]
Thanks! RAM is actually not the exact issue here. So I wanted to train a model with each batch only have rays/cones coming from a single image. Each batch could have only 1% of total rays of a single image, that is, all rays of an image can be divided into 100 consecutive batches. With that in mind,  I tried to simply not shuffle in dataloader. So that the rays are being process in order. But the trained model renders only white images.  

Would you have any thoughts on what might have gone wrong?

**Comment by hjxwhy** [comment]
From my understanding, you want to train the model one by one image, and now you have tried split=="train", batch_type=="all_images", shuffle==False in dataloader, but the model can not converge, yes? How about the loss and training PSNR? If the model seems converged, you can render an image from the training set and an image from val set, making sure the model is overfitting or not.

**Comment by joanncqn** [comment]
[updated]
Yes, that what I did. And the model doesn't seem to converge. I plotted out the loss too, it's simply not converging. 
Fig 1. Loss without shuffle
<img width="845" alt="Screen Shot 2022-07-15 at 11 29 54 AM" src="https://user-images.githubusercontent.com/108494164/179288203-3a737444-c203-437e-8e90-35d182e4e082.png">

I tried to shuffle the rays within an image as well, but seems like it doesn't help much. Any thoughts? Do you think this training method is valid? or perhaps it actually is restraining the model from learning properly? 

Fig 2. Loss with shuffle within an image (train/psnr ~12)
<img width="847" alt="Screen Shot 2022-07-15 at 2 10 39 PM" src="https://user-images.githubusercontent.com/108494164/179311113-0eb3ae91-32a9-4e3f-b70f-abd079e17700.png">

**Comment by hjxwhy** [comment]
Such a learning curve is as invalid as I expected, so random sampling is needed during training. I think the network is learning on one image with gradient descent in one direction, but when this image is finished learning and a new image comes, the gradient descent of this image may not be in the same direction as the last one. So the network will jump repeatedly on a local optimum. Have you made sure random shuffle in all data works well?

**Comment by joanncqn** [comment]
That makes sense. Yeah, when I shuffle all data, it works well and renders image properly.

**Comment by joanncqn** [comment]
Thanks for your input! I'm closing this issue since I was able to fix it by shuffling rays within an image. Really appreciate your help!
